### PR TITLE
Fix number of elements

### PR DESCRIPTION
**Pull request by ezrayst**
Just to fix number of elements, minor change

### CODE REVIEW

File: include/gmsh.tcc [hunk 1]
```diff
@@ -173,8 +173,11 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
   Eigen::VectorXd elementkeyvalues(Tvertices);
   Eigen::VectorXd verticesarray(Tdim * Tvertices);
 
+  const unsigned firstelement = elements_.begin()->first;
+
   //! Iterate through element_
-  for (unsigned i = 0; i < elements_.size(); ++i) {
+  for (unsigned i = 0 + firstelement; i < elements_.size() + firstelement;
```
Comment (kks32) [inline include/gmsh.tcc]: This loop is very confusing. Could you explain the logic behind this loop?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.tcc (hunk 1 above)
Comment (ezrayst) [reply include/gmsh.tcc]: in `gmsh`, the element number does not start from 0 or 1. it started at what is defined here as the "firstelement" (in hexCone, it's 2405 out of some 9000 more elements)

Thus, we cannot loop from 0 here, as the first element ID written in the mesh file is 2405. 

I personally like @cw646's looping from first to last element, but if you want to loop through the element size, we have to offset it this way

Comment (kks32) [reply include/gmsh.tcc]: Why not just use a range-based for loop?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.tcc (hunk 1 above)
Comment (ezrayst) [reply include/gmsh.tcc]: can we do it? `element_` is a map not a vector... we need to get the values of the `first` right?

File: include/gmsh.tcc [hunk 2]
```diff
@@ -174,28 +174,24 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
-      for (unsigned k = 0; k < Tvertices; ++k) {
-        //! Get the vertex wanted from the id
-        auto verticesfind = vertices_.find(elementkeyvalues[k]);
-        //! For each vertex, store the coordinates
-        //! j = 0 -> [X], j = 1 -> [Y], j = 2 -> [Z]
-        for (unsigned l = 0; l < Tdim; ++l) {
-          verticesarray[k * Tdim + l] = verticesfind->second[l];
-        }
+    //! In each element, iterate to get vertices id's of the element
+    for (unsigned j = 0; j < Tvertices; ++j) {
```
Comment (kks32) [inline include/gmsh.tcc]: Do we need this loop? Why not call `element.second[j]`?

File: include/gmsh.tcc [hunk 3]
```diff
@@ -174,28 +174,24 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+      elementkeyvalues[j] = element.second[j];
+    }
+    //! Iterate through the vertices to get coordinates (4 for tetrahedral)
```
Comment (kks32) [inline include/gmsh.tcc]: Aren't we using hexahedron? Why 4?

File: include/gmsh.tcc [hunk 4]
```diff
@@ -174,28 +174,24 @@ void GMSH<Tdim, Tvertices>::store_element_vertices() {
...
+    for (unsigned k = 0; k < Tvertices; ++k) {
+      //! Get the vertex wanted from the id
+      auto verticesfind = vertices_.find(elementkeyvalues[k]);
+      //! For each vertex, store the coordinates
+      //! j = 0 -> [X], j = 1 -> [Y], j = 2 -> [Z]
```
Comment (kks32) [inline include/gmsh.tcc]: There is no `j` in this loop

File: include/gmsh.tcc [hunk 5]
```diff
@@ -211,32 +207,27 @@ void GMSH<Tdim, Tvertices>::compute_material_points() {
...
-      }
+  for (const auto& elementcoord : elementcoordinates_) {
 
-      // Assign the centroid as the coordinate of the material point
-      for (unsigned i = 0; i < Tdim; ++i) {
-        pointsarray[i] = 0;
-        for (unsigned j = 0; j < Tvertices; ++j) {
-          pointsarray[i] += (1. / Tvertices) * m(i, j);
-        }
+    //! Store coordinates in 3x4 matrix
```
Comment (kks32) [inline include/gmsh.tcc]: Need more explanation on what's happening here.

File: include/gmsh.tcc [hunk 6]
```diff
@@ -211,32 +207,27 @@ void GMSH<Tdim, Tvertices>::compute_material_points() {
...
+    Eigen::MatrixXd m(Tdim, Tvertices);
+    for (unsigned i = 0; i < Tvertices; ++i) {
+      for (unsigned j = 0; j < Tdim; ++j) {
+        arrayposition = (i * Tdim) + j;
+        m(j, i) = elementcoord.second[arrayposition];
```
Comment (kks32) [inline include/gmsh.tcc]: Add comments to describe what this is.

File: include/mesh.h [hunk 7]
```diff
@@ -46,14 +46,10 @@ class Mesh {
   }
 
   //! Return a map of mesh element vertices
-  std::map<unsigned, std::array<unsigned, Tdim>> vertices() const {
-    return vertices_;
-  }
+  std::map<unsigned, Eigen::VectorXd> vertices() const { return vertices_; }
```
Comment (kks32) [inline include/mesh.h]: Should we expose these?

File: include/mesh.h [hunk 8]
```diff
@@ -46,14 +46,10 @@ class Mesh {
...
 
   //! Return a map of element id & vertices id
-  std::map<unsigned, std::array<unsigned, Tvertices>> elements() const {
-    return elements_;
-  }
+  std::map<unsigned, Eigen::VectorXd> elements() const { return elements_; }
```
Comment (kks32) [inline include/mesh.h]: Should we need to expose these?

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — APPROVED** [review]

**Pull request merged**